src/nvim/api/vim.c

```
@@ -2389,3 +2389,169 @@ Dictionary nvim_complete_set(Integer index, Dict(complete_set) *opts, Arena *are
+///                 Nvim is waiting for input and would normally not do so.
+///               - clear: Forcefully redraw `win`, all windows containing
+///                 `buf`, or all windows when both are omitted.
+///               - range: Buffer range to redraw, useful for decoration
+///                 providers. Tuple `[first, last]`:
+///                   - first: first line number of the range
+///                   - last: last line number of the range
+///               - cursor: Immediately update cursor position on the screen.
+///               - status: Redraws (a combination of) the following status
+///                 components (array of one or more strings):
+///                 - "bar": |'winbar'|
+///                 - "column": |'statuscolumn'|
+///                 - "line": |'statusline'|
+///                 - "tab": |'tabline'|
+///                 - "all": All of the above
+///               - win: Target a specific |window-ID| for `cursor`, `status`
+///                 and `clear` updates. Current window when 0, all windows if
+///                 omitted and applicable.
+///               - buf: Target a specific buffer number for `range`, `status`,
+///                 and `clear` updates, current buffer when 0 or omitted.
+void nvim_redraw(Dict(redraw) *opts, Error *err)
+  FUNC_API_SINCE(12)
+{
+  win_T *win = NULL;
+  buf_T *buf = NULL;
+  bool save_lz = p_lz;
+  int save_rd = RedrawingDisabled;
+  RedrawingDisabled = 0;
+  p_lz = false;
+
+  if (HAS_KEY(opts, redraw, win)) {
+    win = find_window_by_handle(opts->win, err);
+    if (ERROR_SET(err)) {
+      goto theend;
+    }
+  }
+
+  if (HAS_KEY(opts, redraw, buf)) {
+    buf = find_buffer_by_handle(opts->buf, err);
+    if (ERROR_SET(err)) {
+      goto theend;
+    }
+  }
+
+  if (opts->clear) {
+    if (win != NULL) {
+      redraw_later(win, UPD_NOT_VALID);
+    } else if (buf != NULL) {
+      redraw_buf_later(buf, UPD_NOT_VALID);
```

if we treat `win` and `buf` as mutually exclusive, we should throw an error when both `buf` and `win` are supplied.